src/EVM.hs

```diff
@@ -1542,8 +1548,57 @@ cheatActions =
       action "prank(address)" $
         \sig _ _ input -> case decodeStaticArgs 0 1 input of
           [addr]  -> assign #overrideCaller (Expr.exprToAddr addr)
-          _ -> vmError (BadCheatCode sig)
+          _ -> vmError (BadCheatCode sig),
+
+      action "createFork(string)" $
+        \sig outOffset _ input -> case decodeBuf [AbiStringType] input of
+          CAbi valsArr -> case valsArr of
+            [AbiString bytes] -> do
+              forkId <- length <$> gets (.forks)
+              let urlOrAlias = Char8.unpack bytes
+              modify' $ \vm -> vm { forks = vm.forks Seq.|> ForkState vm.env vm.block vm.cache urlOrAlias }
+              let encoded = encodeAbiValue $ AbiUInt 256 (fromIntegral forkId)
+              assign (#state % #returndata) (ConcreteBuf encoded)
+              copyBytesToMemory (ConcreteBuf encoded) (Lit . num . BS.length $ encoded) (Lit 0) outOffset
+            _ -> vmError (BadCheatCode sig)
+          _ -> vmError (BadCheatCode sig),
 
+      action "selectFork(uint256)" $
+        \sig _ _ input -> case decodeStaticArgs 0 1 input of
+          [forkId] ->
+            forceConcrete forkId "forkId must be concrete" $ \(fromIntegral -> forkId') -> do
+              saved <- Seq.lookup forkId' <$> gets (.forks)
+              case saved of
+                Just forkState -> do
+                  vm <- get
+                  case (vm.env.storage, forkState.env.storage) of
+                    (ConcreteStore store, ConcreteStore savedStore) -> do
+                      let
+                        -- the current contract is persited across forks
+                        self = vm.state.contract
```

Comment:
I don't know exactly how foundry behaves here, but according to the docs only the state of the the test contract itself is persisted between forks, and I think we're going to persist the wrong contract if we call this from a subcontract that's deployed from the top level test contract?